tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -340,12 +338,83 @@ export class WebGPUBackend extends KernelBackend {
     return res;
   }
 
+  private getAndSaveProgram(
+      key: number, getProgram: () => webgpu_program.WebGPUProgram) {
+    if (!this.programCache.has(key)) {
+      this.programCache.set(key, getProgram());
+    }
+    return this.programCache.get(key);
+  }
+
+  // This hash algorithm came from https://source.chromium.org/chromium/chromium/src/+/master:third_party/dawn/src/common/HashUtils.h;l=47;bpv=1;bpt=0
+  private getHash<R extends Rank>(input: Array<boolean|number|ShapeMap[R]|
+      number[]>) {
+    let hash = 0, i, l;
+    const offset = 0x9e3779b9;
+    while (input.length) {
+      // pop value from input
+      const next = input.pop();
+      if (Array.isArray(next)) {
+        for (i = 0, l = next.length; i < l; i++) {
+          hash ^= (hash << 6) + (hash >> 2) + next[i] + offset;
+        }
+        hash ^= (hash << 6) + (hash >> 2) + offset;
+      } else if (typeof(next) === 'number'){
+        if (Number.isInteger(next)){
+          hash ^= (hash << 6) + (hash >> 2) + next + offset;
+        } else {
+          const buffer = new ArrayBuffer(8);
+          const view = new DataView(buffer);
+          view.setFloat64(0, next);
+          const i0 = view.getInt32(0);
+          const i1 = view.getInt32(1);
+          hash ^= (hash << 6) + (hash >> 2) + i0 + i1 + offset;
```

Comment:
Adding the two numbers together like this is collision prone. You should hash-combine each one separately.

Reply:
done.